File: Cosmos-IOS/Cosmostation/Cosmostation/Dao/Price.swift
//
//  Price.swift
//  Cosmostation
//
//  Created by 정용주 on 2021/05/17.
//  Copyright © 2021 wannabit. All rights reserved.
//

import Foundation

public struct Price {
    var denom: String?
    var last_updated: String?
    var prices: Array<Prices> = Array<Prices>()
    
    init(_ dictionary: NSDictionary?) {
        self.denom = dictionary?["denom"] as? String
        self.last_updated = dictionary?["last_updated"] as? String
        self.prices.removeAll()
        if let RawPrices = dictionary?["prices"] as? Array<NSDictionary> {
            for RawPrice in RawPrices {
                self.prices.append(Prices(RawPrice))
            }
        }
    }
    
    public struct Prices {
        var currency: String?
        var current_price: Double?
        var daily_price_change_in_percentage: Double?
        
        init(_ dictionary: NSDictionary?) {
            self.currency = dictionary?["currency"] as? String
            self.current_price = dictionary?["current_price"] as? Double
            self.daily_price_change_in_percentage = dictionary?["daily_price_change_in_percentage"] as? Double
        }
    }
    
    
    public func currencyPrice(_ currency: String) -> NSDecimalNumber {
        if let price = prices.filter({ $0.currency == currency }).first {
            return NSDecimalNumber.init(value: price.current_price!).rounding(accordingToBehavior: WUtils.handler8)
        }
        return NSDecimalNumber.zero.rounding(accordingToBehavior: WUtils.handler8)
    }
    
    public func priceChange() -> NSDecimalNumber {
        if let price = prices.filter({ $0.currency == "usd" }).first {
            return NSDecimalNumber.init(value: price.daily_price_change_in_percentage!).rounding(accordingToBehavior: WUtils.handler2Down)
        }
        return NSDecimalNumber.zero.rounding(accordingToBehavior: WUtils.handler2Down)
    }
}

File: Cosmos-Android/app/src/main/java/wannabit/io/cosmostaion/dao/Price.java
package wannabit.io.cosmostaion.dao;

import com.google.gson.annotations.SerializedName;

import java.math.BigDecimal;
import java.math.RoundingMode;
import java.util.ArrayList;

public class Price {
    @SerializedName("denom")
    public String denom;

    @SerializedName("last_updated")
    public String last_updated;

    @SerializedName("prices")
    public ArrayList<Prices> prices;


    public class Prices {
        @SerializedName("currency")
        public String currency;

        @SerializedName("current_price")
        public Double current_price;

        @SerializedName("daily_price_change_in_percentage")
        public Double daily_price_change_in_percentage;
    }

    public BigDecimal currencyPrice(String currency) {
        for (Prices prices: prices) {
            if (prices.currency == currency) {
                return new BigDecimal(prices.current_price).setScale(8, RoundingMode.DOWN);
            }
        }
        return BigDecimal.ZERO.setScale(8, RoundingMode.DOWN);
    }

    public BigDecimal priceChange() {
        for (Prices prices: prices) {
            if (prices.currency == "usd") {
                return new BigDecimal(prices.daily_price_change_in_percentage).setScale(2, RoundingMode.DOWN);
            }
        }
        return BigDecimal.ZERO.setScale(2, RoundingMode.DOWN);
    }
}
